ctmap: Make GC bpf map dumps more robust.

jrajahalme opened the pull request:
Continue bpf CT map GC iteration from last known found element if the
current element cannot be found. Start again from the beginning
otherwise, but limit restarts from beginning to 2.

Fixes: #3867 
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

borkmann reviewed (approved):
Looks fine to me, although my only concern would be that the restart logic still has a potential to fail. Would be nicer to avoid failures altogether, but the subsequent run would then likely catch these elements later. Should this just be a warn instead of error?

Alternative option could be to get rid of ct_delete*() in the BPF prog and handle everything out of golang protecting the (global) ctmap with a (global) mutex on write ops which currently looks not to be the case. Another one could be to collect all unique keys first via getnextkey and drop the entries in a second run based on the collected keys (failed deletes we can ignore then since they were removed already from elsewhere). But I'm okay with either way, potentially need to revisit once we run into the GC restart error often.

borkmann commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -262,44 +282,74 @@ func doGC6(m *bpf.Map, filter *GCFilter) int {
+
+	if count == m.MapInfo.MaxEntries {
+		// TODO Add a metric we can bump and observe here.
+		log.WithError(err).Warning("Garbage collection on IPv6 CT map failed to finish")
```
Hmm, should that (+ number of iterations) be `count == m.MapInfo.MaxEntries + 1` instead, so we know for sure that there must have been a restart in between? E.g. worst case we could have deleted a full map with all invalid entries, and still trigger this warn, no?

jrajahalme replied:
I had `count` being 0-based and while it was correct, it was misleading, fixed.

borkmann replied:
Ok, thanks. LGTM

jrajahalme committed:
ctmap: Make GC bpf map dumps more robust.

Continue bpf CT map GC iteration from last known found element if the
current element cannot be found. Start again from the beginning
otherwise, but limit overall lookups to maximum number of elements
in the map.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
